File: src/SimpleLevelEditor.Formats.Types/Level/Entity.fs
namespace SimpleLevelEditor.Formats.Types.Level

open System.Numerics
open SimpleLevelEditor.Formats.Types.Level

type Entity =
    { Id: int32 // The id is only used to keep track of the object in the editor.
      mutable Name: string
      mutable Position: Vector3
      mutable Shape: ShapeDescriptor
      mutable Properties: EntityProperty list }

    member this.DeepCopy() =
        { this with
            Shape = this.Shape.DeepCopy()
            Properties = this.Properties |> List.map (_.DeepCopy()) }

    member this.CloneAndPlaceAtPosition(entityId: int32, position: Vector3) =
        { Id = entityId
          Name = this.Name
          Position = position
          Shape = this.Shape.DeepCopy()
          Properties = this.Properties |> List.map (_.DeepCopy()) }

    static member CreateDefault() =
        { Id = 0
          Name = System.String.Empty
          Position = Vector3.Zero
          Shape = ShapeDescriptor.Point
          Properties = [] }

File: src/SimpleLevelEditor.Formats.Level/Entity.cs
using System.Numerics;

namespace SimpleLevelEditor.Formats.Level;

public sealed record Entity
{
	public required int Id { get; init; }

	public required string Name { get; set; }

	public required Vector3 Position { get; set; }

	public required EntityShape Shape { get; set; }

	public required IReadOnlyList<EntityProperty> Properties { get; set; }

	public Entity DeepCopy()
	{
		return this with
		{
			Shape = Shape.DeepCopy(),
			Properties = Properties.Select(p => p.DeepCopy()).ToList(),
		};
	}

	public Entity CloneAndPlaceAtPosition(int entityId, Vector3 position)
	{
		return this with
		{
			Id = entityId,
			Position = position,
			Shape = Shape.DeepCopy(),
			Properties = Properties.Select(p => p.DeepCopy()).ToList(),
		};
	}

	public static Entity CreateDefault()
	{
		return new Entity
		{
			Id = 0,
			Name = string.Empty,
			Position = Vector3.Zero,
			Shape = new EntityShape.Point(),
			Properties = [],
		};
	}
}
